utils/mail.go

```diff
@@ -157,19 +157,18 @@ func sendMail(mimeTo, smtpTo string, from mail.Address, subject, htmlBody string
 		}
 
 		for _, fileInfo := range attachments {
+			bytes, err := fileBackend.ReadFile(fileInfo.Path)
+			if err != nil {
+				return err
+			}
+
 			m.Attach(fileInfo.Name, gomail.SetCopyFunc(func(writer io.Writer) error {
-				bytes, err := fileBackend.ReadFile(fileInfo.Path)
-				if err != nil {
-					return err
-				}
```

Comment:
To be honest, I don't understand why this works. It seems to be a scoping issue. When debugging the code prior to this change, `fileInfo` was always the zeroth element in the `attachments` array when the anonymous function was called by gomail. However, if I read the file before I create the anonymous function and reference `bytes` inside of the anonymous function instead of `fileInfo`, the value is correct. 

I think that this change comes at the cost of increased theoretical memory usage, because all attachments will be held in memory at the same time, as opposed to the original code, in which each file would only be in memory for as long as `bytes` was in scope (i.e. the duration of the anonymous function) before getting cleaned up by the GC.

I'm definitely open to another explanation for what's going on here, and another approach to fixing it, but this one does work.

Reply:
Yeah there's absolutely no way this worked before. But I'm surprised you say `fileInfo` was always the zeroth element. I think it should always be the *last* element:

```go
package main

import (
	"fmt"
)

func main() {
	var funcs []func() string
	for _, x := range []string{"a", "b", "c"} {
		funcs = append(funcs, func() string {
			return x
		})
	}
	for _, x := range funcs {
		fmt.Println(x())
	}
}
```

```
c
c
c
```

– https://play.golang.org/p/Afaj6ms-HqH

This is an example of the one and only "common mistake" on the Go wiki: https://github.com/golang/go/wiki/CommonMistakes

An alternative is to just redefine `fileInfo` inside the loop body with `fileInfo := fileInfo`:

```go
package main

import (
	"fmt"
)

func main() {
	var funcs []func() string
	for _, x := range []string{"a", "b", "c"} {
		x := x
		funcs = append(funcs, func() string {
			return x
		})
	}
	for _, x := range funcs {
		fmt.Println(x())
	}
}
```

```
a
b
c
```

Reply:
Just tested it again, and I was mistaken, `fileInfo` was always the last element for the `attachments` array. 

I read the wiki page that you linked, and I understand what's going on, although I'm surprised that Go doesn't solve for this problem. The compiler is so smart in a lot of other ways, and it seems like it should be able to detect that you're using a variable in a goroutine or an anonymous function and make a copy of it on your behalf.

In any case, thank you for the explanation. I'm going to opt to leave the code as-is, since it works fine, and none of the workarounds on that wiki page would make it any more obvious or readable.

Reply:
Well, I think is hard decision for the compiler, because they don't know if you are sharing the variable on purpose or by accident. You can want to share the variable and the variable changes between all the goroutines, or you can want to isolate that variable from one goroutine to another, that is up to you, and the compiler hasn't a way to know it.

Rust have this kind of protection, but based on the ownership information (which is included in the syntax of the language), but is something that go doesn't have.